### DISCUSSION TITLE
Implement performant instrumentation engine

### DISCUSSION THREAD

**Discussion by MarcoRossignoli** [coverlet-coverage/coverlet#1257] [Ideas]
Today we have some perf issues https://github.com/coverlet-coverage/coverlet/issues?q=is%3Aopen+is%3Aissue+label%3Atenet-performance

But the main one is how the instrumentation is done. We rely heavily on sequences point and we're emitting a lot of code that can lead to a very poor performance during testing(for instance in code with tight loops)

We should try to implement a better strategy or try to fold contiguous hit points when possible.

cc: @daveMueller @petli @tonerdo

**Comment by tonerdo** [comment]
I think we can explore folding, since a single C# line can lead to multiple different instructions that all share they same line numbers. I haven't looked at the instrumentation code in a while but it should be pretty straightforward to perform folding

**Reply by MarcoRossignoli** [reply]
Yep folding will be the first optimization to try for sure.

**Comment by Bertk** [comment]
Closing this discussion since it’s no longer relevant to the current state of the project.
